File: voss/policy.py
"""Signed policy bundles and the default-deny policy engine (Voss RFC 4).

A capability is granted only when a signed policy bundle explicitly
permits the principal, action, and resource within the validity interval.
Missing, malformed, expired, unsigned, or conflicting policy is a denial.
The model and worker never possess the policy signing key.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Sequence, Tuple

from .canonical import ProtocolError, canonical_bytes, loads_strict
from .keys import KeyRing
from .protocol import CanonicalRequest

RUNTIME_MIN_VERSION = "1"
DEFAULT_POLICY_VERSION = "1.0.0"

DECISION_ALLOW = "ALLOW"
DECISION_REQUIRE_APPROVAL = "REQUIRE_APPROVAL"
DECISION_DENY = "DENY"

_ALLOWED_DECISIONS = {DECISION_ALLOW, DECISION_REQUIRE_APPROVAL}
_BODY_FIELDS = {
    "version", "signer", "created_at", "valid_from", "valid_until",
    "min_runtime_version", "audit_required", "revocation_reference", "rules",
}
_RULE_FIELDS = {
    "principal", "action", "resource_prefix", "decision",
    "approval_required", "expiry_limit_seconds",
}


@dataclass(frozen=True)
class Rule:
    principal: str = "*"                       # exact worker principal or "*"
    action: str = ""
    resource_prefix: Optional[Dict[str, Any]] = None
    decision: str = DECISION_ALLOW
    approval_required: bool = False
    expiry_limit_seconds: int = 60


@dataclass
class PolicyBundle:
    version: str
    signer: str
    created_at: str
    valid_from: str
    valid_until: str
    min_runtime_version: str
    audit_required: bool
    revocation_reference: str
    rules: List[Rule] = field(default_factory=list)


class PolicyLoadError(ProtocolError):
    pass


def rule_to_dict(rule: Rule) -> Dict[str, Any]:
    prefix = rule.resource_prefix
    return {
        "principal": rule.principal,
        "action": rule.action,
        "resource_prefix": prefix,
        "decision": rule.decision,
        "approval_required": bool(rule.approval_required),
        "expiry_limit_seconds": int(rule.expiry_limit_seconds),
    }


def build_policy_body(
    rules: Sequence[Rule],
    *,
    version: str = DEFAULT_POLICY_VERSION,
    signer: str = "operator-prototype",
    created_at: Optional[str] = None,
    valid_from: Optional[str] = None,
    valid_until: Optional[str] = None,
    min_runtime_version: str = RUNTIME_MIN_VERSION,
    audit_required: bool = True,
    revocation_reference: str = "",
) -> Dict[str, Any]:
    from datetime import datetime, timedelta, timezone

    now = datetime.now(timezone.utc)
    created_at = created_at or now.isoformat()
    valid_from = valid_from or (now - timedelta(seconds=60)).isoformat()
    valid_until = valid_until or (now + timedelta(hours=1)).isoformat()
    return {
        "version": version,
        "signer": signer,
        "created_at": created_at,
        "valid_from": valid_from,
        "valid_until": valid_until,
        "min_runtime_version": min_runtime_version,
        "audit_required": bool(audit_required),
        "revocation_reference": revocation_reference,
        "rules": [rule_to_dict(r) for r in rules],
    }


def package_policy(body: Dict[str, Any], keyring: KeyRing) -> Dict[str, Any]:
    """Sign the canonical policy body with the Operator key."""
    body_bytes = canonical_bytes(body)
    return {
        "policy": loads_strict(canonical_bytes(body).decode("ascii")),
        "signature": keyring.sign_policy(body_bytes),
    }


def _rule_from_dict(data: Dict[str, Any], index: int) -> Rule:
    unknown = set(data) - _RULE_FIELDS
    if unknown:
        raise PolicyLoadError(f"rule {index}: unknown fields {sorted(unknown)}")
    principal = data.get("principal", "*")
    action = data.get("action")
    if not isinstance(principal, str) or not principal:
        raise PolicyLoadError(f"rule {index}: invalid principal")
    if not isinstance(action, str) or not action:
        raise PolicyLoadError(f"rule {index}: invalid action")
    decision = data.get("decision", DECISION_ALLOW)
    if decision not in _ALLOWED_DECISIONS:
        raise PolicyLoadError(f"rule {index}: unsupported decision {decision!r}")
    approval_required = bool(data.get("approval_required", False))
    expiry = data.get("expiry_limit_seconds", 60)
    if not isinstance(expiry, int) or isinstance(expiry, bool) or not (1 <= expiry <= 86400):
        raise PolicyLoadError(f"rule {index}: invalid expiry_limit_seconds")
    prefix = data.get("resource_prefix")
    if prefix is not None and not isinstance(prefix, dict):
        raise PolicyLoadError(f"rule {index}: resource_prefix must be an object or null")
    if action not in ("workspace.read", "workspace.write", "external.send_mock"):
        # The prototype only ships these action classes; anything else in a
        # signed bundle would represent tooling that does not exist here.
        raise PolicyLoadError(f"rule {index}: action {action!r} is not deployed in this prototype")
    return Rule(
        principal=principal,
        action=action,
        resource_prefix=dict(prefix) if prefix else None,
        decision=decision,
        approval_required=approval_required,
        expiry_limit_seconds=expiry,
    )


class PolicyLoader:
    """Loads, verifies, and performs integrity checks on a policy bundle."""

    def __init__(self, keyring: KeyRing, now: Optional[float] = None):
        self._keyring = keyring
        self._now_func = now or (lambda: __import__("time").time())

    def load_and_verify(self, data: Any) -> PolicyBundle:
        if not isinstance(data, dict):
            raise PolicyLoadError("policy package must be an object")
        body = data.get("policy")
        signature = data.get("signature")
        if not isinstance(body, dict) or not isinstance(signature, str):
            raise PolicyLoadError("policy package must contain body and signature")

        body_bytes = canonical_bytes(body)
        if not self._keyring.verify_policy(body_bytes, signature):
            raise PolicyLoadError("policy signature is invalid")

        unknown = set(body) - _BODY_FIELDS
        if unknown:
            raise PolicyLoadError(f"unknown policy body fields: {sorted(unknown)}")

        version = body.get("version")
        signer = body.get("signer")
        audit_required = body.get("audit_required", True)
        min_runtime = body.get("min_runtime_version")
        rules_data = body.get("rules")

        if not isinstance(version, str) or not version:
            raise PolicyLoadError("policy version missing")
        if not isinstance(signer, str) or not signer:
            raise PolicyLoadError("policy signer missing")
        if audit_required is not True:
            raise PolicyLoadError("policy must require audit (audit_required=False rejected)")
        if min_runtime != RUNTIME_MIN_VERSION:
            raise PolicyLoadError("policy min_runtime_version is not supported")
        if not isinstance(rules_data, list) or not rules_data:
            raise PolicyLoadError("policy must declare at least one rule")

        rules = [_rule_from_dict(r, i) for i, r in enumerate(rules_data)]

        # Human control path: at least one conditional action must exist that
        # requires human approval.  A bundle that removes the last approval
        # gate is rejected (RFC 4.3 / Binding 8.1).
        if not any(r.approval_required or r.decision == DECISION_REQUIRE_APPROVAL for r in rules):
            raise PolicyLoadError("policy removes the last human approval gate")

        now = self._now_func()
        if not self._iso_within(body.get("valid_from"), body.get("valid_until"), now):
            raise PolicyLoadError("policy validity interval does not cover the current time")

        return PolicyBundle(
            version=version,
            signer=signer,
            created_at=body.get("created_at", ""),
            valid_from=body.get("valid_from", ""),
            valid_until=body.get("valid_until", ""),
            min_runtime_version=min_runtime,
            audit_required=True,
            revocation_reference=body.get("revocation_reference", ""),
            rules=rules,
        )

    def _iso_within(self, start: Any, end: Any, now: float) -> bool:
        from datetime import datetime, timezone

        def parse(value: Any) -> Optional[float]:
            if not isinstance(value, str):
                return None
            try:
                dt = datetime.fromisoformat(value)
            except ValueError:
                return None
            if dt.tzinfo is None:
                dt = dt.replace(tzinfo=timezone.utc)
            return dt.timestamp()

        start_ts = parse(start)
        end_ts = parse(end)
        if start_ts is None or end_ts is None:
            return False
        return start_ts <= now <= end_ts

    def now(self) -> float:
        return self._now_func()

    def valid_now(self, bundle: PolicyBundle) -> bool:
        return self._iso_within(bundle.valid_from, bundle.valid_until, self._now_func())


class PolicyEngine:
    """Trusted, deterministic evaluation.  Default-deny."""

    def __init__(self, bundle: PolicyBundle):
        self.bundle = bundle

    @property
    def version(self) -> str:
        return self.bundle.version

    @property
    def valid(self) -> bool:
        return self.bundle is not None

    def evaluate(
        self, cr: CanonicalRequest
    ) -> Tuple[str, str, Optional[Rule]]:
        matches: List[Rule] = []
        for rule in self.bundle.rules:
            if rule.principal not in ("*", cr.principal):
                continue
            if rule.action != cr.action:
                continue
            if not self._matches_resource(rule.resource_prefix, cr.resource):
                continue
            matches.append(rule)

        if not matches:
            return DECISION_DENY, "denied_policy_no_rule", None

        effective = {self._effective(r) for r in matches}
        if len(effective) != 1:
            # Conflicting rules: fail closed (RFC 4.1).
            return DECISION_DENY, "denied_policy_conflict", None

        decision = effective.pop()
        rule = matches[0]
        if decision == DECISION_REQUIRE_APPROVAL:
            return DECISION_REQUIRE_APPROVAL, "policy_approval_required", rule
        return DECISION_ALLOW, "policy_allowed", rule

    def _effective(self, rule: Rule) -> str:
        if rule.approval_required or rule.decision == DECISION_REQUIRE_APPROVAL:
            return DECISION_REQUIRE_APPROVAL
        return DECISION_ALLOW

    def _matches_resource(self, prefix: Optional[Dict[str, Any]], resource: Dict[str, Any]) -> bool:
        if prefix is None:
            return True
        import os

        for key, expected in prefix.items():
            actual = resource.get("path" if key == "path_prefix" else key)
            if key == "path_prefix":
                if not isinstance(actual, str) or not isinstance(expected, str):
                    return False
                expected_norm = os.path.normcase(os.path.realpath(os.path.abspath(expected))).rstrip(os.sep)
                actual_norm = os.path.normcase(actual)
                if not (actual_norm == expected_norm or actual_norm.startswith(expected_norm + os.sep)):
                    return False
            elif actual != expected:
                return False
        return True

File: rust/src/policy.rs
//! Signed policy bundles and the default-deny policy engine.
//!
//! A capability is granted only when a signed policy bundle explicitly
//! permits the principal, action, and resource within the validity interval.
//! The model and worker never possess the policy signing key.

use std::collections::BTreeSet;
use std::time::{SystemTime, UNIX_EPOCH};

use chrono::{DateTime, SecondsFormat, Utc};

use crate::canonical::{Json, ProtocolError, canonical_bytes, loads_strict};
use crate::keys::KeyRing;
use crate::protocol::{CanonicalRequest, path_prefix_matches};

pub const DECISION_ALLOW: &str = "ALLOW";
pub const DECISION_REQUIRE_APPROVAL: &str = "REQUIRE_APPROVAL";
pub const DECISION_DENY: &str = "DENY";

const RUNTIME_MIN_VERSION: &str = "1";
const DEFAULT_POLICY_VERSION: &str = "1.0.0";

const BODY_FIELDS: &[&str] = &[
    "version",
    "signer",
    "created_at",
    "valid_from",
    "valid_until",
    "min_runtime_version",
    "audit_required",
    "revocation_reference",
    "rules",
];

const RULE_FIELDS: &[&str] = &[
    "principal",
    "action",
    "resource_prefix",
    "decision",
    "approval_required",
    "expiry_limit_seconds",
];

const PROTOTYPE_ACTIONS: &[&str] = &["workspace.read", "workspace.write", "external.send_mock"];

#[derive(Clone, Debug, PartialEq)]
pub struct Rule {
    pub principal: String,
    pub action: String,
    pub resource_prefix: Option<Json>,
    pub decision: String,
    pub approval_required: bool,
    pub expiry_limit_seconds: i64,
}

impl Rule {
    pub fn allow(action: impl Into<String>) -> Self {
        Self {
            principal: "*".to_string(),
            action: action.into(),
            resource_prefix: None,
            decision: DECISION_ALLOW.to_string(),
            approval_required: false,
            expiry_limit_seconds: 60,
        }
    }

    pub fn require_approval(action: impl Into<String>) -> Self {
        Self {
            approval_required: true,
            decision: DECISION_REQUIRE_APPROVAL.to_string(),
            ..Self::allow(action)
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct PolicyBundle {
    pub version: String,
    pub signer: String,
    pub created_at: String,
    pub valid_from: String,
    pub valid_until: String,
    pub min_runtime_version: String,
    pub audit_required: bool,
    pub revocation_reference: String,
    pub rules: Vec<Rule>,
}

#[derive(Clone, Debug)]
pub struct PolicyFields {
    pub version: String,
    pub signer: String,
    pub created_at: String,
    pub valid_from: String,
    pub valid_until: String,
    pub min_runtime_version: String,
    pub audit_required: bool,
    pub revocation_reference: String,
}

impl Default for PolicyFields {
    fn default() -> Self {
        let now = Utc::now();
        Self {
            version: DEFAULT_POLICY_VERSION.to_string(),
            signer: "operator-prototype".to_string(),
            created_at: format_iso(now),
            valid_from: format_iso(now - chrono::Duration::seconds(60)),
            valid_until: format_iso(now + chrono::Duration::hours(1)),
            min_runtime_version: RUNTIME_MIN_VERSION.to_string(),
            audit_required: true,
            revocation_reference: String::new(),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct Evaluation {
    pub decision: &'static str,
    pub reason: &'static str,
    pub rule: Option<Rule>,
}

pub fn build_policy_body(rules: &[Rule], fields: &PolicyFields) -> Json {
    Json::object([
        ("version", Json::string(&fields.version)),
        ("signer", Json::string(&fields.signer)),
        ("created_at", Json::string(&fields.created_at)),
        ("valid_from", Json::string(&fields.valid_from)),
        ("valid_until", Json::string(&fields.valid_until)),
        (
            "min_runtime_version",
            Json::string(&fields.min_runtime_version),
        ),
        ("audit_required", Json::Bool(fields.audit_required)),
        (
            "revocation_reference",
            Json::string(&fields.revocation_reference),
        ),
        (
            "rules",
            Json::Array(rules.iter().map(rule_to_json).collect()),
        ),
    ])
}

pub fn package_policy(body: &Json, keyring: &KeyRing) -> Result<Json, ProtocolError> {
    let bytes = canonical_bytes(body)?;
    let signature = keyring.sign_policy(&bytes);
    let normalized = loads_strict(std::str::from_utf8(&bytes).map_err(|_| {
        ProtocolError::new("policy body is not valid ASCII after canonicalization")
    })?)?;
    Ok(Json::object([
        ("policy", normalized),
        ("signature", Json::string(signature)),
    ]))
}

fn rule_to_json(rule: &Rule) -> Json {
    Json::object([
        ("principal", Json::string(&rule.principal)),
        ("action", Json::string(&rule.action)),
        (
            "resource_prefix",
            rule.resource_prefix.clone().unwrap_or(Json::Null),
        ),
        ("decision", Json::string(&rule.decision)),
        ("approval_required", Json::Bool(rule.approval_required)),
        ("expiry_limit_seconds", Json::Int(rule.expiry_limit_seconds)),
    ])
}

pub struct PolicyLoader {
    keyring: KeyRing,
    now: Box<dyn Fn() -> f64 + Send + Sync>,
}

impl PolicyLoader {
    pub fn new(keyring: KeyRing) -> Self {
        Self {
            keyring,
            now: Box::new(system_now),
        }
    }

    pub fn with_clock(keyring: KeyRing, now: impl Fn() -> f64 + Send + Sync + 'static) -> Self {
        Self {
            keyring,
            now: Box::new(now),
        }
    }

    pub fn load_and_verify(&self, data: &Json) -> Result<PolicyBundle, ProtocolError> {
        let package = data
            .as_object()
            .ok_or_else(|| ProtocolError::new("policy package must be an object"))?;
        let body = package
            .get("policy")
            .and_then(Json::as_object)
            .ok_or_else(|| ProtocolError::new("policy package must contain body and signature"))?;
        let signature = package
            .get("signature")
            .and_then(Json::as_str)
            .ok_or_else(|| ProtocolError::new("policy package must contain body and signature"))?;

        let body_json = Json::Object(body.clone());
        let body_bytes = canonical_bytes(&body_json)?;
        if !self.keyring.verify_policy(&body_bytes, signature) {
            return Err(ProtocolError::new("policy signature is invalid"));
        }

        let unknown: Vec<&String> = body
            .keys()
            .filter(|key| !BODY_FIELDS.contains(&key.as_str()))
            .collect();
        if !unknown.is_empty() {
            return Err(ProtocolError::new(format!(
                "unknown policy body fields: {unknown:?}"
            )));
        }

        let version = required_string(body.get("version"), "policy version missing")?;
        let signer = required_string(body.get("signer"), "policy signer missing")?;
        let audit_required = match body.get("audit_required") {
            None => true,
            Some(Json::Bool(true)) => true,
            _ => {
                return Err(ProtocolError::new(
                    "policy must require audit (audit_required=False rejected)",
                ));
            }
        };
        let min_runtime = body.get("min_runtime_version").and_then(Json::as_str);
        if min_runtime != Some(RUNTIME_MIN_VERSION) {
            return Err(ProtocolError::new(
                "policy min_runtime_version is not supported",
            ));
        }
        let rules_data = body
            .get("rules")
            .and_then(|value| match value {
                Json::Array(items) if !items.is_empty() => Some(items),
                _ => None,
            })
            .ok_or_else(|| ProtocolError::new("policy must declare at least one rule"))?;

        let mut rules = Vec::new();
        for (index, rule) in rules_data.iter().enumerate() {
            rules.push(rule_from_json(rule, index)?);
        }
        if !rules
            .iter()
            .any(|rule| rule.approval_required || rule.decision == DECISION_REQUIRE_APPROVAL)
        {
            return Err(ProtocolError::new(
                "policy removes the last human approval gate",
            ));
        }

        let now = (self.now)();
        if !interval_covers(body.get("valid_from"), body.get("valid_until"), now) {
            return Err(ProtocolError::new(
                "policy validity interval does not cover the current time",
            ));
        }

        Ok(PolicyBundle {
            version,
            signer,
            created_at: body
                .get("created_at")
                .and_then(Json::as_str)
                .unwrap_or("")
                .to_string(),
            valid_from: body
                .get("valid_from")
                .and_then(Json::as_str)
                .unwrap_or("")
                .to_string(),
            valid_until: body
                .get("valid_until")
                .and_then(Json::as_str)
                .unwrap_or("")
                .to_string(),
            min_runtime_version: min_runtime.unwrap().to_string(),
            audit_required,
            revocation_reference: body
                .get("revocation_reference")
                .and_then(Json::as_str)
                .unwrap_or("")
                .to_string(),
            rules,
        })
    }

    pub fn load_str(&self, text: &str) -> Result<PolicyBundle, ProtocolError> {
        self.load_and_verify(&loads_strict(text)?)
    }

    pub fn now(&self) -> f64 {
        (self.now)()
    }

    pub fn valid_now(&self, bundle: &PolicyBundle) -> bool {
        let start = Json::string(&bundle.valid_from);
        let end = Json::string(&bundle.valid_until);
        interval_covers(Some(&start), Some(&end), (self.now)())
    }
}

pub struct PolicyEngine {
    pub bundle: PolicyBundle,
}

impl PolicyEngine {
    pub fn new(bundle: PolicyBundle) -> Self {
        Self { bundle }
    }

    pub fn version(&self) -> &str {
        &self.bundle.version
    }

    pub fn valid(&self) -> bool {
        true
    }

    pub fn evaluate(&self, request: &CanonicalRequest) -> Evaluation {
        let mut matches = Vec::new();
        for rule in &self.bundle.rules {
            if rule.principal != "*" && rule.principal != request.principal {
                continue;
            }
            if rule.action != request.action {
                continue;
            }
            if !resource_matches(rule.resource_prefix.as_ref(), &request.resource) {
                continue;
            }
            matches.push(rule);
        }
        if matches.is_empty() {
            return Evaluation {
                decision: DECISION_DENY,
                reason: "denied_policy_no_rule",
                rule: None,
            };
        }
        let effective: BTreeSet<&str> = matches
            .iter()
            .map(|rule| effective_decision(rule))
            .collect();
        if effective.len() != 1 {
            return Evaluation {
                decision: DECISION_DENY,
                reason: "denied_policy_conflict",
                rule: None,
            };
        }
        let decision = *effective.iter().next().expect("one effective decision");
        let rule = matches[0].clone();
        if decision == DECISION_REQUIRE_APPROVAL {
            Evaluation {
                decision: DECISION_REQUIRE_APPROVAL,
                reason: "policy_approval_required",
                rule: Some(rule),
            }
        } else {
            Evaluation {
                decision: DECISION_ALLOW,
                reason: "policy_allowed",
                rule: Some(rule),
            }
        }
    }
}

fn effective_decision(rule: &Rule) -> &'static str {
    if rule.approval_required || rule.decision == DECISION_REQUIRE_APPROVAL {
        DECISION_REQUIRE_APPROVAL
    } else {
        DECISION_ALLOW
    }
}

fn resource_matches(prefix: Option<&Json>, resource: &Json) -> bool {
    let Some(prefix) = prefix else {
        return true;
    };
    let Some(expected_object) = prefix.as_object() else {
        return false;
    };
    let actual_object = resource.as_object();
    for (key, expected) in expected_object {
        if key == "path_prefix" {
            let (Some(expected_path), Some(actual_path)) = (
                expected.as_str(),
                actual_object
                    .and_then(|object| object.get("path"))
                    .and_then(Json::as_str),
            ) else {
                return false;
            };
            if !path_prefix_matches(expected_path, actual_path) {
                return false;
            }
        } else if actual_object.and_then(|object| object.get(key)) != Some(expected) {
            return false;
        }
    }
    true
}

fn rule_from_json(value: &Json, index: usize) -> Result<Rule, ProtocolError> {
    let object = value
        .as_object()
        .ok_or_else(|| ProtocolError::new(format!("rule {index}: invalid principal")))?;
    let unknown: Vec<&String> = object
        .keys()
        .filter(|key| !RULE_FIELDS.contains(&key.as_str()))
        .collect();
    if !unknown.is_empty() {
        return Err(ProtocolError::new(format!(
            "rule {index}: unknown fields {unknown:?}"
        )));
    }
    let principal = object
        .get("principal")
        .and_then(Json::as_str)
        .unwrap_or("*");
    if principal.is_empty() {
        return Err(ProtocolError::new(format!(
            "rule {index}: invalid principal"
        )));
    }
    let action = object.get("action").and_then(Json::as_str).unwrap_or("");
    if action.is_empty() {
        return Err(ProtocolError::new(format!("rule {index}: invalid action")));
    }
    if !PROTOTYPE_ACTIONS.contains(&action) {
        return Err(ProtocolError::new(format!(
            "rule {index}: action {action:?} is not deployed in this prototype"
        )));
    }
    let decision = object
        .get("decision")
        .and_then(Json::as_str)
        .unwrap_or(DECISION_ALLOW);
    if decision != DECISION_ALLOW && decision != DECISION_REQUIRE_APPROVAL {
        return Err(ProtocolError::new(format!(
            "rule {index}: unsupported decision {decision:?}"
        )));
    }
    let approval_required = match object.get("approval_required") {
        None => false,
        Some(Json::Bool(value)) => *value,
        Some(_) => {
            return Err(ProtocolError::new(format!(
                "rule {index}: invalid approval_required"
            )));
        }
    };
    let expiry = match object.get("expiry_limit_seconds") {
        None => 60,
        Some(Json::Int(value)) if (1..=86_400).contains(value) => *value,
        _ => {
            return Err(ProtocolError::new(format!(
                "rule {index}: invalid expiry_limit_seconds"
            )));
        }
    };
    let resource_prefix = match object.get("resource_prefix") {
        None | Some(Json::Null) => None,
        Some(Json::Object(prefix)) => Some(Json::Object(prefix.clone())),
        Some(_) => {
            return Err(ProtocolError::new(format!(
                "rule {index}: resource_prefix must be an object or null"
            )));
        }
    };
    Ok(Rule {
        principal: principal.to_string(),
        action: action.to_string(),
        resource_prefix,
        decision: decision.to_string(),
        approval_required,
        expiry_limit_seconds: expiry,
    })
}

fn required_string(value: Option<&Json>, message: &str) -> Result<String, ProtocolError> {
    match value.and_then(Json::as_str) {
        Some(text) if !text.is_empty() => Ok(text.to_string()),
        _ => Err(ProtocolError::new(message)),
    }
}

fn interval_covers(start: Option<&Json>, end: Option<&Json>, now: f64) -> bool {
    let (Some(start), Some(end)) = (
        start.and_then(Json::as_str).and_then(parse_iso),
        end.and_then(Json::as_str).and_then(parse_iso),
    ) else {
        return false;
    };
    start <= now && now <= end
}

fn parse_iso(value: &str) -> Option<f64> {
    if let Ok(parsed) = DateTime::parse_from_rfc3339(value) {
        return Some(unix_seconds(parsed.with_timezone(&Utc)));
    }
    let naive = chrono::NaiveDateTime::parse_from_str(value, "%Y-%m-%dT%H:%M:%S%.f")
        .or_else(|_| chrono::NaiveDateTime::parse_from_str(value, "%Y-%m-%dT%H:%M:%S"))
        .ok()?;
    Some(unix_seconds(naive.and_utc()))
}

fn unix_seconds(time: DateTime<Utc>) -> f64 {
    time.timestamp() as f64 + f64::from(time.timestamp_subsec_nanos()) / 1_000_000_000.0
}

fn format_iso(time: DateTime<Utc>) -> String {
    time.to_rfc3339_opts(SecondsFormat::Micros, false)
}

fn system_now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_secs_f64())
        .unwrap_or(0.0)
}
